### ISSUE TITLE
How to save the trained model to predict cluster ids of the test set?

### ISSUE THREAD

**Issue by AyberkYavuz** [dstein64/kmeans1d#5]
Let's say we have a one dimensional training set feature that has 10.000 instances.

Also, we have a one dimensional test set feature that has 2.000 instances.

How can we save the trained model in order to load it and produce cluster ids for test set?

```python
import random
import kmeans1d


def generate_items(n_items: int) -> list:
    items = list()
    for i in range(0, n_items):
        n = random.randint(1, 1000)
        items.append(n)
    return items


training_feature = generate_items(10000)

k = 10

clusters, centroids = kmeans1d.cluster(training_feature, k)

# save the trained model

# load the trained model

test_feature = generate_items(2000)

# produce cluster ids for test_feature
```

**Comment by dstein64** [comment]
Hi @AyberkYavuz, Python's built-in `pickle` can be used for saving and load the centroids.

```python
import pickle

...

# save the trained model
with open('/PATH/TO/FILE', 'wb') as f:
    pickle.dump(centroids, f)

# load the trained model
with open('/PATH/TO/FILE', 'rb') as f:
    centroids = pickle.load(f)
```

Cluster IDs for a test set would be the index of the nearest centroid.

```python
# produce cluster ids for test_feature
clusters = []
for x in test_feature:
    distances = [abs(x - c) for c in centroids]
    clusters.append(distances.index(min(distances)))
```

If the Python loops are too slow for your application, using a library would likely be faster.

```python
import numpy as np
from scipy.spatial.distance import cdist
from sklearn.metrics import pairwise_distances_argmin_min

# convert to matrices
test_feature = np.array(test_feature)[:, np.newaxis]
centroids = np.array(centroids)[:, np.newaxis]

# produce cluster ids with numpy
clusters_np = np.argmin(cdist(test_feature, centroids), axis=1)

# produce cluster ids with sklearn
clusters_sklearn = pairwise_distances_argmin_min(test_feature, centroids)[0]
```

**Comment by AyberkYavuz** [comment]
Thank you so much! 😊
